src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+                    var xGetter = input.GetGetter<uint>(_xColIndex);
+                    var yGetter = input.GetGetter<uint>(_yColIndex);
+                    getters[0] = _parent.GetGetter(xGetter, yGetter);
+                }
+                return getters;
+            }
+
+            public IRow GetRow(IRow input, Func<int, bool> predicate, out Action disposer)
+            {
+                var active = Utils.BuildArray(OutputSchema.ColumnCount, predicate);
+                var getters = CreateGetter(input, active);
+                disposer = null;
+                return new SimpleRow(OutputSchema, input, getters);
+            }
+
+            public ISchemaBindableMapper Bindable { get { return _parent; } }
+        }
+
+        public int UserFeaturesSize { get { return -1; } }
+        public KeyType UserIdType { get { return _inputYType; } }
+        public KeyType ItemIdType { get { return _inputXType; } }
+
+        public UserHistoryToItemsMapper GetRecommendMapper(int recommendationCount, bool includeHistory)
+        {
+            _host.CheckParam(recommendationCount > 0, nameof(recommendationCount), "Recommendation count must be positive");
+
+            var scores = new List<KeyValuePair<uint, Float>>(_n);
+            var history = new HashSet<uint>();
+
+            UserHistoryToItemsMapper mapper =
+                (ref uint userId, ref VBuffer<float> userFeatures, ref VBuffer<uint> items, ref VBuffer<float> weights,
+                    ref VBuffer<uint> recoItems, ref VBuffer<float> recoScores) =>
+                {
+                    scores.Clear();
+                    if (!includeHistory)
+                    {
+                        history.Clear();
+                        for (int i = 0; i < items.Count; i++)
+                            history.Add(items.Values[i]);
+                    }
+
+                    int recoCount = Math.Min(_n, recommendationCount);
+                    if (userId == 0)
+                        recoCount = 0;
+                    else
+                    {
+                        int row = (int)userId - 1;
+                        for (int i = 0; i < _n; i++)
+                        {
+                            Single rowScore = Single.NegativeInfinity;
+                            if (includeHistory || !history.Contains((uint)(i + 1)))
+                                rowScore = Score(i, row);
+
+                            scores.Add(new KeyValuePair<uint, float>((uint)i + 1, rowScore));
+                        }
+
+                        scores.Sort((x, y) => y.Value.CompareTo(x.Value));
+                    }
+                    uint[] outItems = Utils.Size(recoItems.Values) < recommendationCount ? new uint[recommendationCount] : recoItems.Values;
+                    Single[] outScores = Utils.Size(recoScores.Values) < recommendationCount ? new Single[recommendationCount] : recoScores.Values;
+
+                    int j;
```

Comment:
>int [](start = 20, length = 3)

nit - move into loop declaration.

Reply:
This interface got removed per Pete's suggestion.

---
In reply to: [225357281](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225357281) [](ancestors = 225357281)